### ISSUE TITLE
Cannot mount child cgroup into LXC container

### ISSUE THREAD

**Issue by evanye** [lxc/lxc#2169]
# Required information

 * Distribution: ubuntu
 * Distribution version: 16.04
 * The output of
   * `lxc-start --version` 2.0.8
   * `lxc-checkconfig`

Kernel configuration not found at /proc/config.gz; searching...
Kernel configuration found at /boot/config-4.4.0-1050-aws
--- Namespaces ---
Namespaces: enabled
Utsname namespace: enabled
Ipc namespace: enabled
Pid namespace: enabled
User namespace: enabled
Warning: newuidmap is not setuid-root
Warning: newgidmap is not setuid-root
Network namespace: enabled
/usr/bin/lxc-checkconfig: 110: [: -lt: unexpected operator
/usr/bin/lxc-checkconfig: 110: [: -eq: unexpected operator

--- Control groups ---
Cgroup: enabled
Cgroup clone_children flag: enabled
Cgroup device: enabled
Cgroup sched: enabled
Cgroup cpu account: enabled
Cgroup memory controller: /usr/bin/lxc-checkconfig: 135: [: -ge: unexpected operator
/usr/bin/lxc-checkconfig: 135: [: -gt: unexpected operator
missing
Cgroup cpuset: enabled

--- Misc ---
Veth pair device: enabled
Macvlan: enabled
Vlan: enabled
Bridges: enabled
Advanced netfilter: enabled
CONFIG_NF_NAT_IPV4: enabled
CONFIG_NF_NAT_IPV6: enabled
CONFIG_IP_NF_TARGET_MASQUERADE: enabled
CONFIG_IP6_NF_TARGET_MASQUERADE: enabled
CONFIG_NETFILTER_XT_TARGET_CHECKSUM: enabled
FUSE (for use with lxcfs): enabled

--- Checkpoint/Restore ---
checkpoint restore: enabled
CONFIG_FHANDLE: enabled
CONFIG_EVENTFD: enabled
CONFIG_EPOLL: enabled
CONFIG_UNIX_DIAG: enabled
CONFIG_INET_DIAG: enabled
CONFIG_PACKET_DIAG: enabled
CONFIG_NETLINK_DIAG: enabled
File capabilities: /usr/bin/lxc-checkconfig: 170: [: -gt: unexpected operator

Note : Before booting a new kernel, you can check its configuration
usage : CONFIG=/path/to/config /usr/bin/lxc-checkconfig

   * `uname -a`
Linux ip-10-89-188-186 4.4.0-1050-aws #59-Ubuntu SMP Tue Jan 30 19:57:10 UTC 2018 x86_64 x86_64 x86_64 GNU/Linux

   * `cat /proc/self/cgroup`

ubuntu@ip-10-89-188-186:~$ cat /proc/self/cgroup
11:hugetlb:/
10:blkio:/user.slice
9:devices:/user.slice
8:cpuset:/
7:pids:/user.slice/user-1000.slice
6:cpu,cpuacct:/user.slice
5:net_cls,net_prio:/
4:freezer:/user/ubuntu/0
3:memory:/user/ubuntu/0
2:perf_event:/
1:name=systemd:/user.slice/user-1000.slice/session-16.scope

   * `cat /proc/1/mounts`

ubuntu@ip-10-89-188-186:~$ cat /proc/self/cgroup
11:hugetlb:/
10:blkio:/user.slice
9:devices:/user.slice
8:cpuset:/
7:pids:/user.slice/user-1000.slice
6:cpu,cpuacct:/user.slice
5:net_cls,net_prio:/
4:freezer:/user/ubuntu/0
3:memory:/user/ubuntu/0
2:perf_event:/
1:name=systemd:/user.slice/user-1000.slice/session-16.scope
ubuntu@ip-10-89-188-186:~$ cat /proc/1/mounts
sysfs /sys sysfs rw,nosuid,nodev,noexec,relatime 0 0
proc /proc proc rw,nosuid,nodev,noexec,relatime 0 0
udev /dev devtmpfs rw,nosuid,relatime,size=7691660k,nr_inodes=1922915,mode=755 0 0
devpts /dev/pts devpts rw,nosuid,noexec,relatime,gid=5,mode=620,ptmxmode=000 0 0
tmpfs /run tmpfs rw,nosuid,noexec,relatime,size=1539928k,mode=755 0 0
/dev/xvda1 / ext4 rw,relatime,discard,data=ordered 0 0
securityfs /sys/kernel/security securityfs rw,nosuid,nodev,noexec,relatime 0 0
tmpfs /dev/shm tmpfs rw,nosuid,nodev 0 0
tmpfs /run/lock tmpfs rw,nosuid,nodev,noexec,relatime,size=5120k 0 0
tmpfs /sys/fs/cgroup tmpfs ro,nosuid,nodev,noexec,mode=755 0 0
cgroup /sys/fs/cgroup/systemd cgroup rw,nosuid,nodev,noexec,relatime,xattr,release_agent=/lib/systemd/systemd-cgroups-agent,name=systemd 0 0
pstore /sys/fs/pstore pstore rw,nosuid,nodev,noexec,relatime 0 0
cgroup /sys/fs/cgroup/perf_event cgroup rw,nosuid,nodev,noexec,relatime,perf_event 0 0
cgroup /sys/fs/cgroup/memory cgroup rw,nosuid,nodev,noexec,relatime,memory 0 0
cgroup /sys/fs/cgroup/freezer cgroup rw,nosuid,nodev,noexec,relatime,freezer 0 0
cgroup /sys/fs/cgroup/net_cls,net_prio cgroup rw,nosuid,nodev,noexec,relatime,net_cls,net_prio 0 0
cgroup /sys/fs/cgroup/cpu,cpuacct cgroup rw,nosuid,nodev,noexec,relatime,cpu,cpuacct 0 0
cgroup /sys/fs/cgroup/pids cgroup rw,nosuid,nodev,noexec,relatime,pids 0 0
cgroup /sys/fs/cgroup/cpuset cgroup rw,nosuid,nodev,noexec,relatime,cpuset,clone_children 0 0
cgroup /sys/fs/cgroup/devices cgroup rw,nosuid,nodev,noexec,relatime,devices 0 0
cgroup /sys/fs/cgroup/blkio cgroup rw,nosuid,nodev,noexec,relatime,blkio 0 0
cgroup /sys/fs/cgroup/hugetlb cgroup rw,nosuid,nodev,noexec,relatime,hugetlb 0 0
systemd-1 /proc/sys/fs/binfmt_misc autofs rw,relatime,fd=29,pgrp=1,timeout=0,minproto=5,maxproto=5,direct 0 0
debugfs /sys/kernel/debug debugfs rw,relatime 0 0
hugetlbfs /dev/hugepages hugetlbfs rw,relatime 0 0
mqueue /dev/mqueue mqueue rw,relatime 0 0
fusectl /sys/fs/fuse/connections fusectl rw,relatime 0 0
lxcfs /var/lib/lxcfs fuse.lxcfs rw,nosuid,nodev,relatime,user_id=0,group_id=0,allow_other 0 0
tmpfs /run/user/1000 tmpfs rw,nosuid,nodev,relatime,size=1539928k,mode=700,uid=1000,gid=1000 0 0
/dev/mapper/vg-lv /local_disk0 ext4 rw,relatime,data=ordered 0 0
/dev/xvdb /var/lib/lxc ext4 rw,relatime,data=ordered 0 0

# Issue description

I'm trying to allow processes inside the container from modifying a child cgroup that is assigned to the container. However, by creating the child cgroup, I materialize the parent cgroup, and LXC avoids using existing cgroups (see https://github.com/lxc/lxc/blob/master/src/lxc/cgroups/cgfsng.c#L1678), so the container cgroup will not be a parent of the child cgroup.

# Steps to reproduce

 1. Step one
add a `lxc.mount.entry` to the lxc config like
```
lxc.mount.entry = /sys/fs/cgroup/memory/lxc/CONTAINER/container_child mnt/cgroup_memory default bind,create=dir 0 0
```

 2. Step two
Observe that the memory cgroups have all been created, e.g.:
- `/sys/fs/cgroup/memory/lxc/CONTAINER/container_child/...`
- `/sys/fs/cgroup/memory/lxc/CONTAINER/...`

 3. Step three
`lxc-start -n CONTAINER`
When the container starts up, it notices that the CONTAINER memory cgroup already exists, so that it instead creates a cgroup at `/sys/fs/cgroup/memory/lxc/CONTAINER-1/`. This unfortunately makes it so that the child cgroup does not inherit the parent's cgroup information

# Information to attach

 - [ ] any relevant kernel output (`dmesg`)
 - [ ] container log (The <log> file from running `lxc-start -n <c> -l <log> -o DEBUG`)
 - [ ] the containers configuration file

**Comment by brauner** [comment]
So you want two containers to be in the same cgroup?

**Comment by evanye** [comment]
That's one possible use case, not quite the one I was intending. I'll try describing it more clearly:

On my host machine, I run a LXC container for other users. Since the users cannot be trusted, I set cgroup limits (specifically memory/cpu) so that a OOM in the container will not affect my services on the host.

On top of this, I would like to allow users on the guest to be able to set up cgroups for guest processes. For example, the guest OS will set up multiple processes that would like to be in separate cgroups.

The way I tried this was to give the LXC container it's own cgroup (call is CONTAINER), and create one child cgroup (call it CONTAINER/container_child), then **bind mount** the child cgroup directory into the container so that the container has access to read/update the child cgroup.

This runs into a cyclic dependency - as far as I know, in order to bind mount a directory into a container, the directory structure must be set up before the container has started. However, if the cgroup directory structure exists, then starting the LXC container will have the container use a separate cgroup, e.g. CONTAINER-1, so that the child cgroups (CONTAINER/container_child) does NOT inherit any cgroup limits set on the container.

If I try to flip this around, I can start the container, which creates the correct cgroup (CONTAINER), however it looks like we cannot dynamically bind mount the sub-cgroup (CONTAINER/container_child) into the running container.
